platforms/core-configuration/graph-serialization/src/main/kotlin/org/gradle/internal/serialize/graph/Codec.kt

```diff
@@ -60,6 +60,9 @@ interface WriteContext : IsolateContext, MutableIsolateContext, Encoder {
 }
 
 
+interface CloseableWriteContext : WriteContext, AutoCloseable
```

Comment:
💭  Some thoughts on naming

When I introduced `CloseableServiceRegistry`, Lóránt commented that this name indicates that the class has a `close` method, but it does not tell what responsibility users have when they get an instance of this class. I think it's a fair point that applies here as well.

In other words, it's unclear whether users are supposed to ensure they close the instance, once they have used it. It's also unclear whether the functions that returns `CloseableWriterContext` are designed around one or the other expectation.

Making a parallel with the Kotlin Coroutines, they call these things *scopes*, and they make it [very](https://github.com/Kotlin/kotlinx.coroutines/blob/08156771dd2142d74fda063a31e4c4cddb674a9c/kotlinx-coroutines-core/common/src/CoroutineScope.kt#L17) [clear](https://github.com/Kotlin/kotlinx.coroutines/blob/08156771dd2142d74fda063a31e4c4cddb674a9c/kotlinx-coroutines-core/common/src/CoroutineScope.kt#L49) in the documentation that the user has the responsibility to close them. They go further to ensure good DX, and introduce `coroutineScope {}` functions that take care of creating and *closing* the scope.

I think we should do something similar here. We should probably find another name for `CloseableWriteContext`, but I am not sure something like `WriteContextScope` is the way to go.

Reply:
> but it does not tell what responsibility users have when they get an instance of this class

I find that argument a little esoteric to be honest.

Reply:
The missing information on the expectations of who needs to close the context is very practical. Misinterpreting the expectation leads to either double-closing or, worse, leaking resources.

Reply:
The missing information can be perfectly addressed by a documentation comment without the need for a new naming convention, would you agree?

Reply:
The documentation would definitely help, and we should strive to always have it.

Having a name that signals the responsibility is always better, though. But you are right that this might be not only the name question, but also the convention question. At the same time, I don't think we have a convention for this pattern in the Gradle codebase. Such that when an engineer sees, say, `CloseableAbc`, it's understood they need to treat the instance as a resource. At least I am not aware of it.